apps/src/feedback.js

```diff
@@ -708,21 +714,28 @@ FeedbackUtils.prototype.getShareFailure_ = function(options) {
 /**
  * Generates an appropriate feedback message
  * The message will be one of the following, from highest to lowest precedence:
- * 0. Failure override message specified on level (options.level.failureMessageOverride)
- * 1. Message passed in by caller (options.message).
- * 2. Header message due to dashboard text check fail (options.response.share_failure).
- * 3. Level-specific message (e.g., options.level.emptyBlocksErrorMsg) for
+ * 1. End of lesson message for last level in a lesson.
+ * 2. Failure override message specified on level
+ * (options.level.failureMessageOverride)
+ * 2. Message passed in by caller (options.message).
+ * 3. Header message due to dashboard text check fail (options.response.share_failure).
+ * 4. Level-specific message (e.g., options.level.emptyBlocksErrorMsg) for
  *    specific result type (e.g., TestResults.EMPTY_BLOCK_FAIL).
- * 4. System-wide message (e.g., msg.emptyBlocksErrorMsg()) for specific
+ * 5. System-wide message (e.g., msg.emptyBlocksErrorMsg()) for specific
  *    result type (e.g., TestResults.EMPTY_BLOCK_FAIL).
  * @param {FeedbackOptions} options
  * @return {string} message
  */
 FeedbackUtils.prototype.getFeedbackMessage = function(options) {
   var message;
 
-  // If a message was explicitly passed in, use that.
   if (
+    options.level &&
+    options.level.lastLevelInLesson &&
+    options.level.showEndOfLessonMsgs
+  ) {
+    message = msg.endOfLesson();
```

Comment:
Do we only want to show this message when the student has passed the level?

(Caveat: I don't totally understand where/how this function is used.  It seems unexpected to me that all of the cases above would be handled in the same method.)

Reply:
Chatted with Product and Curriculum and we're ok with showing this message regardless of whether a student has passed or completed the level. CSD and CSP don't have validated levels at the end of lessons, so it's not an issue. CSA will have validated levels at the end of lessons, but they aren't in the system yet and will cross that bridge when they come to it if they want additional customization of this logic. 

The issue you've raised become a much bigger consideration if we extend this "end of lesson" experience to CSF, but I'd hope by that point we'll have time to holistically refactor and consolidate the finish/challenge dialogs. 🤞

Reply:
Ok, reshuffled the logic a bit to handle end of lesson updates for both validated and unvalidated scenarios.